update ethereum and universal provider, add RPC reference section (#1485)
WalletConnect/walletconnect-docs: docs/advanced/providers/ethereum.mdx

import Table from '../../components/Table'

# Ethereum Provider

[EIP-1193](https://eips.ethereum.org/EIPS/eip-1193) compliant Provider for the WalletConnect v2 protocol.
Built on top of [Universal Provider](./universal.mdx) and [WalletConnectModal](../walletconnectmodal/about.mdx).
You can use this on it's own or pass it down to libraries like ethers, viem or web3js.

## Installation

```bash npm2yarn
npm install @walletconnect/ethereum-provider
```

## Initialization

Initialize Ethereum Provider by calling its `init` method and passing down the required arguments:

```ts
import { EthereumProvider } from '@walletconnect/ethereum-provider'

const provider = await EthereumProvider.init({
  projectId: 'YOUR_PROJECT_ID',
  metadata: {
    name: 'My Website',
    description: 'My Website Description',
    url: 'https://mywebsite.com', // origin must match your domain & subdomain
    icons: ['https://avatars.githubusercontent.com/u/37784886']
  },
  showQrModal: true,
  optionalChains: [1, 137, 2020],

/*Optional - Add custom RPCs for each supported chain*/
  rpcMap: {
    1: 'mainnet.rpc...',
    137: 'polygon.rpc...'
  },
})
```
:::info

> - Make sure that the `url` from the `metadata` matches your domain and subdomain. This will later be used by the [Verify API](../../cloud/verify.md) to tell wallets if your application has been verified or not.

Reviewer: ```suggestion
- Make sure that `url` from `metadata` matches your domain and subdomain. This will later be used by [Verify API](../../cloud/verify.md) to confirm if your application has been verified or not.
```

- We recommend using `optionalChains` (optional namespaces) over `chains` (required namespaces).  
_Required namespaces_ will block wallets from connecting to your application if any of the chains are not supported by the wallet. Smart Contract Wallets can only support one chain, 
the one that they had been deployed to, this might cause issues when using required namespaces.
- Using `chains`, `methods` or `events` will create a _Required Namespaces_ object internally.
- `optionalMethods` and `optionalChains` default to the following methods and events: [Read source code](https://github.com/WalletConnect/walletconnect-monorepo/blob/v2.0/providers/ethereum-provider/src/constants/rpc.ts). 

> - If `rpcMap` is not defined it will fallback into the [Blockchain API](../../cloud/blockchain-api.md) RPCs. Keep in mind that the Blockchain API supports a limited [list of chains](https://github.com/WalletConnect/rpc-proxy/blob/master/SUPPORTED_CHAINS.md).

Reviewer: ```suggestion
- If `rpcMap` is not defined it will fallback to [Blockchain API](../../cloud/blockchain-api.md) RPCs. Keep in mind that Blockchain API supports a limited [list of chains](https://github.com/WalletConnect/rpc-proxy/blob/master/SUPPORTED_CHAINS.md).
```

      type: { code: 'Metadata' },
      required: 'false'
    },
    {
      value: 'qrModalOptions',
      description: 'An array of WalletConnectModal options. See https://docs.walletconnect.com/web3modal/options',
      type: { code: 'QrModalOptions' },
      required: 'false'
    },
    {
      value: 'chains',
      description: 'An array of required chain IDs you want to support. If the wallet does not support this chains it will not be able to connect. Not recommended for multi-chain applications.',
      type: { code: 'number[]' },
      required: 'false'
    },
    {
      value: 'methods',
      description: 'The required methods you want to support.  Not recommended for multi-chain applications.',
      type: { code: 'string[]' },
      required: 'false'
    },
    {
      value: 'events',
      description: 'The required events you want to support. Not recommended for multi-chain applications.',
      type: { code: 'string[]' },
      required: 'false'
    },
  ]}
/>

## Use with WalletConnectModal

> When `showQrModal` is enabled, ethereum provider will automatically show and hide [WalletConnectModal](../walletconnectmodal/about.mdx).

Reviewer: ```suggestion
When `showQrModal` is enabled, EthereumProvider will automatically show and hide [WalletConnectModal](../walletconnectmodal/about.mdx).
```
